src/config/store.js

```diff
@@ -0,0 +1,83 @@
+import {createStore, applyMiddleware, compose} from 'redux';
+import createSagaMiddleware from 'redux-saga';
+
+import rootReducer from '../reducers';
+import rootSaga from '../sagas';
+import {APPLICATION_LOCAL_STORAGE_KEY} from './config';
+import {initialState as gifInitialState} from '../reducers/gif';
+
+// Middlewares
+const sagaMiddleware = createSagaMiddleware();
+const middlewares = [
+  sagaMiddleware
+];
+
+// Enhancers
+const createEnhancers = () => {
+  let enhancers = [];
+
+  /**
+   * Uncomment this lines and add the redux-devtools package to enable redux dev
+   *  tools into development mode.
+   */
+  // if (process.env.NODE_ENV !== 'production') {
+  //   enhancers.push(
+  //     window.__REDUX_DEVTOOLS_EXTENSION__
+  //       && window.__REDUX_DEVTOOLS_EXTENSION__());
+  // }
+
+  return enhancers;
+};
+
+const persistStateDetails = (state) => {
+  try {
+    let stateCopy = JSON.parse(JSON.stringify(state));
+
+    // Remove unnecesary state details
+    stateCopy.gifs.fetching = gifInitialState.fetching;
+    stateCopy.gifs.displayed = gifInitialState.displayed;
+    stateCopy.gifs.favoritesTextFilter = gifInitialState.favoritesTextFilter;
+
+    let serializedState = JSON.stringify(stateCopy);
+    localStorage.setItem(APPLICATION_LOCAL_STORAGE_KEY, serializedState);
+  } catch (e) {}
+}
+
+const retrievePreviousStateDetails = () => {
+  let result = {};
+
+    try {
+      let serializedState =
+        window.localStorage.getItem(APPLICATION_LOCAL_STORAGE_KEY);
+
+      if (serializedState) {
+        result = JSON.parse(serializedState);
+      }
+    } catch (e) {}
+
+  return result;
+}
+
+const enhancers = createEnhancers();
+const composedEnhancers = (enhancers.length > 0)
+  ? compose(applyMiddleware(...middlewares), ...enhancers)
+  : applyMiddleware(...middlewares);
+
+
+const initializeStore = (initialState) => {
+  if (!initialState) {
+    // Rehydrate the store
+    initialState = retrievePreviousStateDetails();
+  }
+
+  // Store object
+  const store = createStore(rootReducer, initialState, composedEnhancers);
+
+  store.subscribe(() => persistStateDetails(store.getState()));
```

Comment:
This is a good solution, but keep in mind that this is going to be executed every time the store changes. When dealing with bigger apps (where serializing the whole state could be more expensive in terms of performance), you could only hear for specific actions with a saga or middleware to avoid serializing on each change (in this case, you could only backup the state when a favorite is added or removed). But for the purposes of this exercise, this is enough 👍